lucene/spatial-extras/src/java/org/apache/lucene/spatial/prefix/tree/S2PrefixTreeCell.java

```diff
@@ -0,0 +1,285 @@
+            children =  FACES;
+        }
+        else {
+            children = new S2CellId[4];
+            children[0] = cellId.childBegin();
+            children[1] = children[0].next();
+            children[2] = children[1].next();
+            children[3] = children[2].next();
+        }
+        List<Cell> cells = new ArrayList<>(children.length);
+        for (S2CellId pixel : children) {
+            cells.add(new S2PrefixTreeCell(tree, pixel));
+        }
+        return new FilterCellIterator(cells.iterator(), shapeFilter);
+    }
+
+    @Override
+    public Shape getShape(){
+        if (shape==null){
+            if (cellId == null) { //World cell
+                return tree.getSpatialContext().getWorldBounds();
+            }
+            return tree.s2ShapeFactory.getS2CellShape(cellId);
+        }
+        return shape;
+    }
+
+    @Override
+    public boolean isPrefixOf(Cell c) {
+        if (cellId == null) {
+            return true;
+        }
+        S2PrefixTreeCell cell =(S2PrefixTreeCell)c;
+        return cellId.contains(cell.cellId);
+    }
+
+    @Override
+    public int compareToNoLeaf(Cell fromCell) {
+        if (cellId == null) {
+            return 1;
+        }
+        S2PrefixTreeCell cell = (S2PrefixTreeCell)fromCell;
+        return cellId.compareTo(cell.cellId);
+    }
+
+    /**
+     * Check if a cell is a leaf.
+     *
+     * @param ref The Byteref of the leaf
+     * @return true if it is a leaf, e.g last byte is the special Character.
+     */
+    private boolean isLeaf(BytesRef ref){
+        return (ref.bytes[ref.offset + ref.length - 1] == LEAF);
+    }
+
+    /**
+     * Get the {@link S2CellId} from the {@link BytesRef} representation.
+     *
+     * @param ref The bytes.
+     * @return the corresponding S2 cell.
+     */
+    private S2CellId getS2CellIdFromBytesRef(BytesRef ref){
+        int length = ref.length;
+        if (isLeaf(ref)){
+            length--;
+        }
+        if (length == 0) {
+            return null; //world cell
+        }
+        int face = PIXELS.get(ref.bytes[ref.offset]);
+        S2CellId cellId = FACES[face];
+        //we will do it directly with cell ids for performance
+        long id = cellId.id();
+        for (int i=ref.offset+1; i<ref.offset + length; i++){
+            int pos = PIXELS.get(ref.bytes[i]);
+            long oldLsb = id & -id;
+            id  = id - oldLsb + (oldLsb >>> 2);
+            id  = id + pos * ((id & -id) << 1);
+        }
+        return new S2CellId(id);
+    }
+
+    /**
+     * Codify a {@link S2CellId} into its {@link BytesRef} representation.
+     *
+     * @param cellId The S2 Cell id to codify.
+     * @param bref The byteref representation.
+     */
+    private void getBytesRefFromS2CellId(S2CellId cellId, BytesRef bref){
+        if (cellId == null) {//world cell
+            bref.length=0;
+            return;
+        }
+        int length = getLevel() + 1;
+        byte[] b = new byte[length];
```

Comment:
the bref parameter may already have bytes that is of this length.  If it is, reuse it thus avoiding object allocation.  BytesRef exists to avoid object allocation and copying.